com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs

```diff
@@ -491,6 +546,329 @@ private void OnValidate()
+                        removeEmptyPrefabs.Add(i);
+                    }
+                }
+
+                // If we have a player prefab, then we need to verify it is in the list of NetworkPrefabOverrideLinks for client side spawning.
+                if (NetworkConfig.PlayerPrefab != null)
+                {
+                    var playerPrefabNetworkObject = NetworkConfig.PlayerPrefab.GetComponent<NetworkObject>();
+                    if (playerPrefabNetworkObject != null)
+                    {
+                        //In the event there is no NetworkPrefab entry (i.e. no override for default player prefab)
+                        if (!NetworkConfig.NetworkPrefabOverrideLinks.ContainsKey(playerPrefabNetworkObject
+                            .GlobalObjectIdHash))
+                        {
+                            //Then add a new entry for the player prefab
+                            var playerNetworkPrefab = new NetworkPrefab();
+                            playerNetworkPrefab.Prefab = NetworkConfig.PlayerPrefab;
+                            NetworkConfig.NetworkPrefabs.Insert(0, playerNetworkPrefab);
+                            NetworkConfig.NetworkPrefabOverrideLinks.Add(playerPrefabNetworkObject.GlobalObjectIdHash,
+                                playerNetworkPrefab);
+                        }
+                    }
+                    else
+                    {
+                        // Provide the name of the prefab with issues so the user can more easily find the prefab and fix it
+                        Debug.LogError($"{nameof(NetworkConfig.PlayerPrefab)} (\"{NetworkConfig.PlayerPrefab.name}\") has no NetworkObject assigned to it!.");
+                    }
+                }
+
+                // Clear out anything that is invalid or not used (for invalid entries we already logged warnings to the user earlier)
+                // Iterate backwards so indices don't shift as we remove
+                for (int i = removeEmptyPrefabs.Count - 1; i >= 0; i--)
+                {
+                    NetworkConfig.NetworkPrefabs.RemoveAt(removeEmptyPrefabs[i]);
+                }
+
+                removeEmptyPrefabs.Clear();
+
+                if (!IsServer)
+                {
+                    State = NetworkManagerState.AwaitingApproval;
+                    if (!NetworkConfig.NetworkTransport.StartClient())
+                    {
+                        ShutdownInternal();
+                        InvokeOnStartupFailedCallback(StartupFailureReason.ConnectFailed);
+                        return false;
+                    }
+                }
+                else
+                {
+                    if (NetworkConfig.NetworkTransport.StartServer())
+                    {
+                        State = NetworkManagerState.Ready;
+                        InvokeOnReadyCallback();
```

Comment:
I am confused as to what this callback is for?
For the server-host it is invoked after the Transport is started, but a few lines down OnServerStarted is invoked.  So, for server-host it seems like it signifies that the transport has been started?
For the client it is invoked only if scene management is disabled and happens just before the InvokeOnClientConnectedCallback is invoked too... so if scene management is enabled then InvokeOnReadyCallback never is invoked for the client.

This action seems like it is close to the same thing for OnServerStarted with the exception being that it is invoked when the transport is started successfully and is pretty much the same thing as InvokeOnClientConnectedCallback with the exception that it only gets invoked when scene management is disabled (is this intentional and perhaps comments as to why).

Perhaps some additional comments as to what InvokeOnReadyCallback is supposed to signify would help?

Reply:
We talked about this a little the other day, but just to put it in writing...

When addressables are in use, anything that needs them can't be used or spawned until after they've been loaded. Additionally, we've had issues in the past in general with people trying to do things before all the startup had been completed and it not working. This callback is one of two ways the user can know when the SDK is fully ready to use (connected, all addressable assets loaded, and for clients, connection has been approved by the server). The other way is for them to poll the State value in an Update event and watch for it to be Ready. Either way, it gives the user something they've been missing - an easy and reliable way to know when the SDK is ready.

With addressables, assets are loaded before the transport connection starts, which means that there's even more things that can't be done immediately after the return from StartClient()/StartHost()/StartServer(). That increases the need for the user to have access to this kind of information - that's why I chose now to add the state machine and callbacks, because the SDK with addressables would be very frustrating and difficult to use without this.

You are right that this tends to be invoked at the same time as OnServerStarted and OnClientConnected, but it's generic and the naming is clearer, to me. `OnClientConnected` is overloaded - sometimes it means "I successfully connected", other times it means "someone connected to me", making it more confusing to implement because you have to treat it differently between client and server, and you also have to know that Host will be treated like a server and will get OnServerStarted instead of OnClientConnected.

It's true the user could use those two callbacks to detect the same event, but I'm a fan of things that aren't confusing, and I this callback can also be reused in both client and server code, which the others can't. If I could, I'd actually remove the other two, but since I can't do that in 1.0, I can at least add a clearer and easier-to-use alternative.